docs/architecture/grpc-for-wcf-developers/migrate-duplex-services.md

```diff
@@ -0,0 +1,481 @@
+{
+    public IStockPriceSubscriber GetSubscriber(string[] symbols)
+    {
+        return new StockPriceSubscriber(symbols);
+    }
+}
+```
+
+#### Registering the factory
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddGrpc();
+    services.AddSingleton<IStockPriceSubscriberFactory, StockPriceSubscriberFactory>();
+}
+```
+
+Now, this class can be used to implement the gRPC StockTicker service.
+
+#### StockTickerService.cs
+
+```csharp
+public class StockTickerService : Protos.SimpleStockTicker.SimpleStockTickerBase
+{
+    private readonly IStockPriceSubscriberFactory _subscriberFactory;
+
+    public StockTickerService(IStockPriceSubscriberFactory subscriberFactory)
+    {
+        _subscriberFactory = subscriberFactory;
+    }
+
+    public override async Task Subscribe(SubscribeRequest request, IServerStreamWriter<StockTickerUpdate> responseStream, ServerCallContext context)
+    {
+        var subscriber = _subscriberFactory.GetSubscriber(request.Symbols.ToArray());
+
+        subscriber.Update += async (sender, args) =>
+            await WriteUpdateAsync(responseStream, args.Symbol, args.Price);
+
+        await AwaitCancellation(context.CancellationToken);
+    }
+
+    private async Task WriteUpdateAsync(IServerStreamWriter<StockTickerUpdate> stream, string symbol, decimal price)
+    {
+        try
+        {
+            await stream.WriteAsync(new StockTickerUpdate
+            {
+                Symbol = symbol,
+                PriceCents = (int)(price * 100),
+                Time = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow)
+            });
+        }
+        catch (Exception e)
+        {
+            // Handle any errors due to broken connection etc.
+            _logger.LogError($"Failed to write message: {e.Message}");
+        }
+    }
+
+    private static Task AwaitCancellation(CancellationToken token)
+    {
+        var completion = new TaskCompletionSource<object>();
+        token.Register(() => completion.SetResult(null));
+        return completion.Task;
+    }
+}
+```
+
+As you can see, although the declaration in the `.proto` file says the method "returns" a stream of `StockTickerUpdate` messages, in fact it returns a vanilla `Task`. The job of creating the stream is handled by the generated code and the gRPC runtime libraries, which provide the `IServerStreamWriter<StockTickerUpdate>` response stream, ready to use.
+
+Unlike a WCF duplex service, where the instance of the service class is kept alive while the connection is open, the gRPC service uses the returned Task to keep the service alive. The Task shouldn't complete until the connection is closed.
+
+The service can tell when the client has closed the connection by using the `CancellationToken` from the `ServerCallContext`. A simple static method, `AwaitCancellation`, is used to create a Task that completes when the token is canceled.
```

Comment:
A better way for the client to tell the server to end in bidirectional calls is the client completing the RequestStream (RequestStream.CompleteAsync). I prefer it over a cancellation token.

IMO it is better because it allows the server and client to gracefully complete. A cancellation token is a big "STOP RIGHT NOW" hammer, and results in errors being thrown in various places (in Kestrel, in HttpClient, in gRPC client).

There are places where cancellation token approach is useful (killing a hanging call, or server streaming where there is no other option to indicate to the server to stop) but I think for a regular request to stop, being graceful is better.